File: server/lib/src/features/books/Filters.ts
import { BookStatus } from "./Book";

export class Filters {
  query?: string
  tags: string[] = []
  genres: string[] = []
  statuses: string[] = [
    BookStatus.abandoned, 
    BookStatus.completed, 
    BookStatus.inProgress
  ]

  static fromQuery(query: any, forProfile: string | undefined) {
    const { query: queryStr, tags, genres, statuses } = query;
    
    var filters: Filters = new Filters();

    if (query) {
      filters.query = queryStr as string;
    }

    if (tags) {
      filters.tags = (Array.isArray(tags) ? tags : [tags]) as string[];
    }

    if (genres) {
      filters.genres = (Array.isArray(genres) ? genres : [genres]) as string[];
    }

    if (statuses) {
      filters.statuses = (Array.isArray(statuses) ? statuses : [statuses]) as string[];
    }
    return filters
  }
}

File: client/lib/src/features/books/domain/filters.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'book.dart';

part 'filters.freezed.dart';
part 'filters.g.dart';

@freezed
class Filters with _$Filters {
  factory Filters({
    @JsonKey(includeIfNull: false) String? ofProfile,
    @JsonKey(includeIfNull: false) String? forProfile,
    @JsonKey(includeIfNull: false) String? query,
    @Default([]) List<String> tags,
    @Default([]) List<String> genres,
    @Default([]) List<BookStatus> statuses
  }) = _Filters;

  factory Filters.fromJson(Map<String, dynamic> json) => _$FiltersFromJson(json);
}